Merge/engelsystem b5ce5bd.

kobaj opened the pull request:
Lots of commits here:

1. Vast majority are just merging from https://github.com/engelsystem/engelsystem 
    *  (so, you know, when you go to merge, don't squash this or we lose all of that context)
1. Then a bunch of fixes for things that broke from the merge
1. Finally started to update how the navigation and permissions should work

I recommend you pull down this branch and run `./startup.sh` to see the changes yourself! But here are some screenshots

![image](https://github.com/user-attachments/assets/449bc9fe-a497-46a3-ace6-3533aacf55fe)
![image](https://github.com/user-attachments/assets/1a861e69-815d-4040-9219-8e482cc99960)
![image](https://github.com/user-attachments/assets/d3dbbfe1-1c54-4f8d-8722-dfcf0252a78d)

chipuni commented on startup.sh:
```diff
@@ -6,6 +6,16 @@ if [ $? -ne 0 ]; then
+compose() {
+  if command -v docker-compose >/dev/null 2>&1; then
+    # Use docker-compose if it's installed
+    docker-compose "$@"
```
I wrote the following test in `/bin/sh`

```
 foobar() {
    echo "$0"
}
```

Then I typed `foobar build --no-cache`

It did not echo `build --no-cache`.

kobaj replied:
Did you mean to type `$0` (dollarsign zero), or `$@` (dollarsign atsign) like my code? 

https://linuxhandbook.com/bash-special-variables/

chipuni replied:
Aha. I mistyped. That explains why my test wasn't working. Thank you!

chipuni reviewed (changes requested):
I don't have time to go over the new code in detail.

When I tried running it locally, it immediately crashed. Here is the output of `docker logs engelsystem-es_server-1`:

```
Starting the nginx server.
Beginning migration.
Exception: Code: 0, Message: Error: Unable to connect to database, File: src/Database/DatabaseServiceProvider.php:67, Previous: SQLSTATE[HY000] [2002] Connection refused, Trace: [{"file":"\/var\/www\/src\/Database\/DatabaseServiceProvider.php","line":44,"function":"exitOnError","class":"Engelsystem\\Database\\DatabaseServiceProvider","type":"->","args":[{"errorInfo":["HY000",2002,"Connection refused"]}]},{"file":"\/var\/www\/src\/Application.php","line":62,"function":"register","class":"Engelsystem\\Database\\DatabaseServiceProvider","type":"->","args":[]},{"file":"\/var\/www\/src\/Application.php","line":84,"function":"register","class":"Engelsystem\\Application","type":"->","args":[{}]},{"file":"\/var\/www\/includes\/application.php","line":14,"function":"bootstrap","class":"Engelsystem\\Application","type":"->","args":[{"providers":["Engelsystem\\Logger\\LoggerServiceProvider","Engelsystem\\Exceptions\\ExceptionsServiceProvider","Engelsystem\\Config\\ConfigServiceProvider","Engelsystem\\Helpers\\ConfigureEnvironmentServiceProvider","Engelsystem\\Events\\EventsServiceProvider","Engelsystem\\Http\\UrlGeneratorServiceProvider","Engelsystem\\Renderer\\RendererServiceProvider","Engelsystem\\Database\\DatabaseServiceProvider","Engelsystem\\Http\\RequestServiceProvider","Engelsystem\\Http\\SessionServiceProvider","Engelsystem\\Helpers\\Translation\\TranslationServiceProvider","Engelsystem\\Http\\ResponseServiceProvider","Engelsystem\\Http\\Psr7ServiceProvider","Engelsystem\\Helpers\\AuthenticatorServiceProvider","Engelsystem\\Helpers\\AssetsServiceProvider","Engelsystem\\Renderer\\TwigServiceProvider","Engelsystem\\Middleware\\RouteDispatcherServiceProvider","Engelsystem\\Middleware\\RequestHandlerServiceProvider","Engelsystem\\Middleware\\SessionHandlerServiceProvider","Engelsystem\\Http\\Validation\\ValidationServiceProvider","Engelsystem\\Http\\RedirectServiceProvider","Engelsystem\\Helpers\\VersionServiceProvider","Engelsystem\\Mail\\MailerServiceProvider","Engelsystem\\Http\\HttpClientServiceProvider","Engelsystem\\Helpers\\DumpServerServiceProvider","Engelsystem\\Helpers\\UuidServiceProvider"],"middleware":["Engelsystem\\Middleware\\SendResponseHandler","Engelsystem\\Middleware\\ExceptionHandler","Engelsystem\\Middleware\\SetLocale","Engelsystem\\Middleware\\ETagHandler","Engelsystem\\Middleware\\AddHeaders","Engelsystem\\Middleware\\TrimInput","Engelsystem\\Middleware\\ErrorHandler","Engelsystem\\Middleware\\VerifyCsrfToken","Engelsystem\\Middleware\\RouteDispatcher","Engelsystem\\Middleware\\SessionHandler","Engelsystem\\Middleware\\RequestHandler"],"event-handlers":{"message.created":"Engelsystem\\Events\\Listener\\Messages@created","news.created":"Engelsystem\\Events\\Listener\\News@created","oauth2.login":"Engelsystem\\Events\\Listener\\OAuth2@login","shift.entry.deleting":["Engelsystem\\Events\\Listener\\Shift@deletedEntryCreateWorklog","Engelsystem\\Events\\Listener\\Shift@deletedEntrySendEmail"]}}]},{"file":"\/var\/www\/bin\/migrate","line":13,"args":["\/var\/www\/includes\/application.php"],"function":"require_once"}]
<pre style="background-color:#333;color:#ccc;z-index:1000;position:fixed;bottom:1em;padding:1em;width:97%;max-height:90%;overflow-y:auto;">Exception: (0)
array(3) {
  [&quot;string&quot;]=&gt;
  string(36) &quot;Error: Unable to connect to database&quot;
  [&quot;file&quot;]=&gt;
  string(43) &quot;src/Database/DatabaseServiceProvider.php:67&quot;
  [&quot;stacktrace&quot;]=&gt;
  array(5) {
    [0]=&gt;
    array(2) {
      [&quot;file&quot;]=&gt;
      string(14) &quot;bin/migrate:13&quot;
      [&quot;require_once&quot;]=&gt;
      array(1) {
        [0]=&gt;
        string(33) &quot;/var/www/includes/application.php&quot;
      }
    }
    [1]=&gt;
    array(2) {
      [&quot;file&quot;]=&gt;
      string(27) &quot;includes/application.php:14&quot;
      [&quot;bootstrap&quot;]=&gt;
      array(1) {
        [0]=&gt;
        string(25) &quot;Engelsystem\Config\Config&quot;
      }
    }
    [2]=&gt;
    array(2) {
      [&quot;file&quot;]=&gt;
      string(22) &quot;src/Application.php:84&quot;
      [&quot;register&quot;]=&gt;
      array(1) {
        [0]=&gt;
        string(44) &quot;Engelsystem\Database\DatabaseServiceProvider&quot;
      }
    }
    [3]=&gt;
    array(2) {
      [&quot;file&quot;]=&gt;
      string(22) &quot;src/Application.php:62&quot;
      [&quot;register&quot;]=&gt;
      array(0) {
      }
    }
    [4]=&gt;
    array(2) {
      [&quot;file&quot;]=&gt;
      string(43) &quot;src/Database/DatabaseServiceProvider.php:44&quot;
      [&quot;exitOnError&quot;]=&gt;
      array(1) {
        [0]=&gt;
        string(12) &quot;PDOException&quot;
      }
    }
  }
}
</pre>
```

(Also see below about the code...)

kobaj commented:
Thanks for the log printout! I think this is because the database is taking a while to load. I replaced the 5 second wait I added with a true health check. Let me know if that works or still has issues. :)

chipuni commented:
My system is (obviously) not set up with ConCat. So it errored when I pressed the button "Login using Concat". Let me know if you want a copy of that log. (You do NOT need to fix this problem.)

chipuni commented:
You will need to set up values for the following fields in `resources/lang/en_US/default.po`: ( I will be updating this list as I come across examples.)

`footer.eventinfo.name_start_end `

chipuni commented:
The output for date/times looks like

`4America/Los_Angeles12America/Los_AngelesMon, 25 Dec 2023 16:42:04 -0800pmMonday.25pm31America/Los_Angeles314212America/Los_Angeles`

I fixed this problem in another version; this might be part of the fix: https://github.com/chipuni/engelsystem/commit/df847c1e4f1d5061e525381682cb50001389a8e5

This problem occurs in many parts of the UI.

chipuni commented:
You will need to read through the files in `resources/lang/en_US` and make at least the following replacements in the `msgstr` lines:

angel --> gofur

(I didn't see anything for 'heaven' that was another change I needed to make.)

For example, 

```
msgid "shift.angeltype_source"
msgstr "Needed angels from: %s"
```

needs to change to

```
msgid "shift.angeltype_source"
msgstr "Needed gofurs from: %s"
```

kobaj commented:
> My system is (obviously) not set up with ConCat. So it errored when I pressed the button "Login using Concat". Let me know if you want a copy of that log. (You do NOT need to fix this problem.)

Fixed! 

> You will need to set up values for the following fields in resources/lang/en_US/default.po: ( I will be updating this list as I come across examples.)
> footer.eventinfo.name_start_end

Good catch! Fixed!

> The output for date/times looks like...

I'm not sure where you are seeing these malformed dates? Is there a particular page or location? For example, what I see: 
[image](https://github.com/user-attachments/assets/d1107578-6283-4674-8289-8fe0561f9c74)

> You will need to read through the files in resources/lang/en_US

I thought I hit them all, but you're right, I missed a few. Here is a helpful regex I made to make sure I got them all this time: `perl -0777 -pe "s/msgid.*?msgstr//gsm" ./resources/lang/en_US/default.po | grep -i "angel"`

chipuni commented:
Here is an example of a screen with both a bad date AND the bad footer. This is from the latest version of Engelsystem.

<img width="1756" alt="Bad date and bad footer" src="https://github.com/user-attachments/assets/bfd9416f-8f42-4ac7-8cad-aa493ae7cd30" />

For the footer, you will need to edit the file `resources/lang/en_US/default.po` and add the following lines:

```
msgid "footer.eventinfo.name_start_end"
msgstr "%1$s, from %2$s until %3$s"
```

Here's another example with the malformed dates:

<img width="1756" alt="Screenshot 2024-12-19 at 14 44 13" src="https://github.com/user-attachments/assets/310b6b27-c87c-4d3f-804f-a14499159dbc" />

chipuni commented:
But the doc for angels looks good.

kobaj commented:
> msgid "footer.eventinfo.name_start_end"

Ooooh! I see where I got mixed up last time. I fixed `footer.eventinfo.start_end` but I also needed to fix `footer.eventinfo.**NAME**start_end` too. :)

> Here's another example with the malformed dates:

Excellent! Thank you so much, with that screenshot I was able to track down and fix the problem.
